File: 238.product-of-array-except-self.java
// @leet start

class Solution {
  public int[] productExceptSelf(int[] nums) {
    int[] left = new int[nums.length];
    int[] right = new int[nums.length];
    right[0] = 1;
    left[0] = 1;
    int lefty = 1;
    int righty = 1;
    for (int i = 0; i < nums.length; i++) {
      right[nums.length - i - 1] = righty;
      left[i] = lefty;
      lefty *= nums[i];
      righty *= nums[nums.length - i - 1];
    }
    int[] answer = new int[nums.length];
    for (int i = 0; i < nums.length; i++) {
      answer[i] = left[i] * right[i];
    }
    return answer;
  }
}
// @leet end

File: 238.product-of-array-except-self.py
# @leet start
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = []
        right = []
        lefty = 1
        righty = 1
        for i in range(0, len(nums)):
            left.append(lefty)
            right.append(righty)
            lefty *= nums[i]
            righty *= nums[len(nums) - i - 1]

        answer = []
        for i in range(0, len(nums)):
            answer.append(left[i] * right[len(nums) - i - 1])
        return answer


# @leet end
